README.md
LDAP support for the XP Framework
========================================================================

[![Build status on GitHub](https://github.com/xp-framework/ldap/workflows/Tests/badge.svg)](https://github.com/xp-framework/ldap/actions)
[![XP Framework Module](https://raw.githubusercontent.com/xp-framework/web/master/static/xp-framework-badge.png)](https://github.com/xp-framework/core)
[![BSD Licence](https://raw.githubusercontent.com/xp-framework/web/master/static/licence-bsd.png)](https://github.com/xp-framework/core/blob/master/LICENCE.md)
[![Requires PHP 7.0+](https://raw.githubusercontent.com/xp-framework/web/master/static/php-7_0plus.svg)](http://php.net/)
[![Supports PHP 8.0+](https://raw.githubusercontent.com/xp-framework/web/master/static/php-8_0plus.svg)](http://php.net/)
[![Latest Stable Version](https://poser.pugx.org/xp-framework/ldap/version.png)](https://packagist.org/packages/xp-framework/ldap)

The peer.ldap package implements LDAP (Lighweight Directory Access Protocol) access.

Example (LDAP search)
---------------------

```php
use peer\ldap\LDAPConnection;
use util\cmd\Console;

$l= new LDAPConnection('ldap://ldap.example.com');
$l->connect();

$search= $l->search(
  'ou=People,dc=OpenLDAP,dc=Org', 
  '(objectClass=*)'
);
  
Console::writeLinef('===> %d entries found', $search->numEntries());
foreach ($search as $result) {
  Console::writeLine('---> ', $result->toString());
}

$l->close();
```

Example (Modifying an entry)
----------------------------

```php
use peer\ldap\{LDAPConnection, LDAPEntry};

$l= new LDAPConnection('ldap://uid=admin,o=roles,dc=planet-xp,dc=net:password@ldap.example.com');
$l->connect();

with ($entry= $l->read(new LDAPEntry('uid=1549,o=people,dc=planet-xp,dc=net'))); {
  $entry->setAttribute('firstname', 'Timm');

  $l->modify($entry);
}

$l->close();
```

Example (Adding an entry)
-------------------------

```php
use peer\ldap\{LDAPConnection, LDAPEntry};

$l= new LDAPConnection('ldap://uid=admin,o=roles,dc=planet-xp,dc=net:password@ldap.example.com');
$l->connect();

with ($entry= new LDAPEntry('uid=1549,o=people,dc=planet-xp,dc=net')); {
  $entry->setAttribute('uid', 1549);
  $entry->setAttribute('firstname', 'Timm');
  $entry->setAttribute('lastname', 'Friebe');
  $entry->setAttribute('objectClass', 'xpPerson');

  $l->add($entry);
}

$l->close();
```

Dynamically creating LDAP queries
---------------------------------
If the LDAP queries need to be constructed dynamically the LDAPQuery
class provides a printf-style syntax to do so:

```php
use peer\ldap\LDAPQuery;

$res= $ldap->searchBy(new LDAPQuery(
  'o=people,dc=planet-xp,dc=net',
  '(&(objectClass=%c)(|(username=%s)(uid=%d)))',
  'xpPerson',
  'friebe'
  1549
));
```

When using the "%s" token, the value passed is escaped according to 
rules in LDAP query syntax. The %c token copies as-is, and %d handles
the argument as numeric value.


ChangeLog.md
LDAP protocol support for the XP Framework ChangeLog
========================================================================

## ?.?.? / ????-??-??

## 9.1.1 / 2022-11-02

* Fixed issue #15: BC break between 9.0 and 9.1 when iterating over
  LDAPSearchResult
  (@thekid)

## 9.1.0 / 2022-02-27

* Made `LDAPSearchResult` implement `lang.Closeable` - @thekid
* Made library compatible with XP 11 - @thekid
* Simplified iteration code in `LDAPSearchResult` - @thekid

## 9.0.0 / 2020-04-11

* Implemented xp-framework/rfc#334: Drop PHP 5.6:
  . **Heads up:** Minimum required PHP version now is PHP 7.0.0
  . Rewrote code base, grouping use statements
  . Rewrote `isset(X) ? X : default` to `X ?? default`
  . Converted `newinstance` to anonymous classes
  (@thekid)

## 8.1.3 / 2020-04-04

* Made compatible with XP 10 - @thekid

## 8.1.2 / 2018-10-26

* Merged PR #14: Bug: Use correct order of calls - @johannes85

## 8.1.1 / 2018-07-30

* Fixed issue #13: Warning *"continue" targeting switch is equivalent to 
  "break"* which started occurring in PHP 7.3
  (@thekid)

## 8.1.0 / 2018-06-06

* Added `peer.ldap.LDAPNoSuchObject` exception for `LDAP_NO_SUCH_OBJECT`
  errors. This is useful when performing searches on subtrees that exist
  in 99.9% of all cases, so we don't check them beforehand.
  (@thekid)

## 8.0.1 / 2018-04-02

* Fixed compatiblity with PHP 7.2 - @thekid

## 8.0.0 / 2017-06-15

* **Heads up:** Dropped PHP 5.5 support - now requires PHP 5.6 minimum!
  (@thekid)
* Fixed issue #9: *Function ldap_sort() is deprecated* by implementing
  sorting in userland. See PR #19
  (@thekid)
* **Heads up**: Removed deprecated peer.ldap.LDAPClient class. Use the
  new peer.ldap.LDAPConnection class introduced with PR #5 in April 2016
  (@thekid)
* Added forward compatibility with XP 9.0.0 - @thekid

## 7.4.3 / 2017-05-02

* Fixed "lang.NullPointerException (Undefined variable: user)" - @thekid

## 7.4.2 / 2016-09-17

* Fixed "undefined variable attrs" in `LDAPConnection::search()` - @thekid

## 7.4.1 / 2016-08-29

* Made compatible with xp-framework/network v8.0.0 - @thekid

## 7.4.0 / 2016-08-28

* Added forward compatibility with XP 8.0.0 - @thekid

## 7.3.0 / 2016-07-03

* Merged PR #8: Detect disconnects - @thekid
* Extended `connect()` to optionally accept DN and password with which
  to bind for instead of the credentials passed in the connection DSN.
  (@thekid)
* Rewrote search() and searchBy() internally without the need for
  `call_user_func_array()`.
  (@thekid)

## 7.2.0 / 2016-05-02

* Merged PR #6: Add accessor for underlying connection DSN - @thekid

## 7.1.1 / 2016-04-22

* Fixed LDAP options not being correctly set in `connect()` - @thekid

## 7.1.0 / 2016-04-19

* Implemented xp-framework/rfc#147: new LDAP API. See pull request #5
  (@thekid)

## 7.0.0 / 2016-02-22

* **Adopted semantic versioning. See xp-framework/rfc#300** - @thekid 
* Added version compatibility with XP 7 - @thekid
* **Heads up**: Changed minimum XP version to XP 6.5.0, and with it the
  minimum PHP version to PHP 5.5.
  (@thekid)

## 6.2.0 / 2016-01-23

* **Heads up**: Change minimum XP version required to *6.3.0*.
  (@thekid)
* Fix code to use `nameof()` instead of the deprecated `getClassName()`
  method from lang.Generic. See xp-framework/core#120
  (@thekid)

## 6.1.1 / 2015-08-24

* Fixed `LDAPClient::read()` which was broken after refactoring
  (@thekid)

## 6.1.0 / 2015-08-23

* **Heads up**: Changed LDAPSearchResult's getFirstEntry(), getEntry() &
  getNextEntry() methods to return `NULL` instead of `FALSE` when EOF is
  reached. While technically this is a BC break, it does not break the
  iteration advertised in the README file. See xp-framework/ldap#4.
  (@thekid)
* Implemented xp-framework/ldap#3: LDAP entries iteration via `foreach`
  (@thekid)

## 6.0.2 / 2015-02-12

* Changed dependency to use XP ~6.0 (instead of dev-master) - @thekid

## 6.0.1 / 2015-01-16

* Merged PR #1: Add class constants for SCOPE (@kiesel)

## 6.0.0 / 2015-01-10

* Removed superfluous encoding and decoding, XP encoding is utf-8 as is
  the encoding for all LDAP calls.
  (@thekid)
* Heads up: Converted classes to PHP 5.3 namespaces - (@thekid)


src/main/php/peer/ldap/util/BerStream.class.php
<?php namespace peer\ldap\util;

use io\streams\BufferedInputStream;
use io\streams\InputStream;
use io\streams\OutputStream;

class BerStream {
  const EOC = 0;
  const BOOLEAN = 1;
  const INTEGER = 2;
  const BITSTRING = 3;
  const OCTETSTRING = 4;
  const NULL = 5;
  const OID = 6;
  const OBJECTDESCRIPTOR = 7;
  const EXTERNAL = 8;
  const REAL = 9;
  const ENUMERATION = 10;
  const PDV = 11;
  const UTF8STRING = 12;
  const RELATIVEOID = 13;
  const SEQUENCE = 16;
  const SET = 17;
  const NUMERICSTRING = 18;
  const PRINTABLESTRING = 19;
  const T61STRING = 20;
  const VIDEOTEXSTRING = 21;
  const IA5STRING = 22;
  const UTCTIME = 23;
  const GENERALIZEDTIME = 24;
  const GRAPHICSTRING = 25;
  const VISIBLESTRING = 26;
  const GENERALSTRING = 28;
  const UNIVERSALSTRING = 29;
  const CHARACTERSTRING = 30;
  const BMPSTRING = 31;
  const CONSTRUCTOR = 32;
  const CONTEXT = 128;

  const SEQ_CTOR = 48;   // (SEQUENCE | CONSTRUCTOR)

  protected $write= [''];
  protected $read= [0];
  protected $buffer= null;

  protected $in, $out;

  /**
   * Constructor
   *
   * @param  io.streams.InputStream $in
   * @param  io.streams.OutputStream $out
   */
  public function __construct(InputStream $in, OutputStream $out) {

    // Debug
    // $in= newinstance('io.streams.InputStream', [$in], [
    //   'backing'     => null,
    //   '__construct' => function($backing) { $this->backing= $backing; },
    //   'read'        => function($length= 8192) {
    //     $chunk= $this->backing->read($length);
    //     if (null !== $chunk) {
    //       \util\cmd\Console::writeLine(BerStream::dump($chunk, '<<<'));
    //     }
    //     return $chunk;
    //   },
    //   'available'   => function() { return $this->backing->available(); },
    //   'close'       => function() { $this->backing->close(); }
    // ]);
    // $out= newinstance('io.streams.OutputStream', [$out], [
    //   'backing'     => null,
    //   '__construct' => function($backing) { $this->backing= $backing; },
    //   'write'       => function($chunk) {
    //     \util\cmd\Console::writeLine(BerStream::dump($chunk, '>>>'));
    //     return $this->backing->write($chunk);
    //   },
    //   'flush'       => function() { return $this->backing->flush(); },
    //   'close'       => function() { $this->backing->close(); }
    // ]);

    $this->in= $in instanceof BufferedInputStream ? $in : new BufferedInputStream($in, 8192);
    $this->out= $out;
  }

  private static function chars($bytes, $start, $offset) {
    $s= '';
    for ($j= $start; $j < min($offset, strlen($bytes)); $j++) {
      $c= $bytes{$j};
      $s.= ($c < "\x20" || $c > "\x7F" ? '.' : $c);
    }
    return $s;
  }

  public static function dump($bytes, $message= null) {
    $n= strlen($bytes);
    $next= ' ';
    if (null === $message) {
      $s= '';
      $p= 74;
    } else {
      $s= $message.' ';
      $p= 74 - strlen($message) - 1;
    }
    $s.= str_pad('== ('.$n." bytes) ==\n", $p, '=', STR_PAD_LEFT);
    $o= 0;
    while ($o < $n) {
      $s.= sprintf('%04x: ', $o);
      for ($i= 0; $i < 16; $i++) {  
        if ($i + $o >= $n) {
          $s.= 7 === $i ? '    ' :  '   ';
        } else {
          $s.= sprintf('%02x %s', ord($bytes{$i + $o}), 7 === $i ? ' ' : '');
        }
      }
      $o+= $i;
      $s.= '|'.str_pad(self::chars($bytes, $o - 16, $o), 16, ' ', STR_PAD_RIGHT)."|\n";
    }
    return $s;
  }

  /**
   * Starts writing a sequence
   *
   * @param  int $tag
   */
  public function startSequence($tag= self::SEQ_CTOR) {
    array_unshift($this->write, '');
    $this->writeByte($tag);
  }

  /**
   * Encode length
   *
   * @param  int $l
   * @return string encoded bytes
   * @throws lang.IllegalStateException if length is > 0xffffff
   */
  protected function encodeLength($l) {
    if ($l <= 0x7f) {
      return pack('C', $l);
    } else if ($l <= 0xff) {
      return "\x81".pack('C', $l);
    } else if ($l <= 0xffff) {
      return "\x82".pack('CC', $l >> 8, $l);
    } else if ($l <= 0xffffff) {
      return "\x83".pack('CCC', $l << 16, $l >> 8, $l);
    } else {
      throw new \lang\IllegalStateException('Length too long: '.$l);
    }
  }

  /**
   * Write raw bytes to current sequence
   *
   * @param  string $raw
   */
  public function write($raw) {
    $this->write[0].= $raw;
  }

  /**
   * Write single byte to current sequence
   *
   * @param  int $b
   */
  public function writeByte($b) {
    $this->write[0].= pack('C', $b);
  }

  /**
   * Write length to current sequence
   *
   * @param  int $l
   */
  public function writeLength($l) {
    $this->write[0].= $this->encodeLength($l);
  }

  /**
   * Write integer to current sequence
   *
   * @param  int $i
   * @param  int $tag
   */
  public function writeInt($i, $tag= self::INTEGER) {
    if ($i < -0xffffff || $i >= 0xffffff) {
      $len= 4;
    } else if ($i < -0xffff || $i >= 0xffff) {
      $len= 3;
    } else if ($i < -0xff || $i >= 0xff) {
      $len= 2;
    } else {
      $len= 1;
    }
    $this->write[0].= pack('CC', $tag, $len).substr(pack('N', $i), -$len);
  }

  /**
   * Write NULL value to current sequence
   *
   * @param  string $raw
   */
  public function writeNull() {
    $this->write[0].= pack('C', self::NULL)."\x00";
  }

  /**
   * Write boolean to current sequence
   *
   * @param  bool $b
   * @param  int $tag
   */
  public function writeBoolean($b, $tag= self::BOOLEAN) {
    $this->write[0].= pack('C', $tag)."\x01".($b ? "\xff" : "\x00");
  }

  /**
   * Write string to current sequence
   *
   * @param  string $s
   * @param  int $tag
   */
  public function writeString($s, $tag= self::OCTETSTRING) {
    $length= $this->encodeLength(strlen($s));
    $this->write[0].= pack('C', $tag).$length.$s;
  }

  /**
   * Write enumeration to current sequence
   *
   * @param  int $e
   * @param  int $tag
   */
  public function writeEnumeration($e, $tag= self::ENUMERATION) {
    $this->writeInt($e, $tag);
  }

  /**
   * Write enumeration to current sequence
   *
   * @param  string $buffer
   * @param  int $tag
   */
  public function writeBuffer($buffer, $tag) {
    $this->writeByte($tag);
    $this->writeLength(strlen($buffer));
    $this->write($buffer);
  }

  /**
   * Writes a filter
   *
   * @param  peer.ldap.filter.Filter
   * @return void
   */
  public function writeFilter($filter) {
    $this->{'writeFilter'.$filter->kind}($filter);
  }

  /** FILTER_AND = 0xa0 */
  private function writeFilterAnd($and) {
    $this->startSequence(0xa0);
    foreach ($and->filters as $filter) {
      $this->{'writeFilter'.$filter->kind}($filter);
    }
    $this->endSequence();
  }

  /** FILTER_AND = 0xa1 */
  private function writeFilterOr($or) {
    $this->startSequence(0xa1);
    foreach ($or->filters as $filter) {
      $this->{'writeFilter'.$filter->kind}($filter);
    }
    $this->endSequence();
  }

  /** FILTER_AND = 0xa2 */
  private function writeFilterNot($not) {
    $this->startSequence(0xa2);
    $this->{'writeFilter'.$not->filter->kind}($not->filter);
    $this->endSequence();
  }

  /** FILTER_EQUALITY = 0xa3 */
  private function writeFilterEquality($eq) {
    $this->startSequence(0xa3);
    $this->writeString($eq->attribute);
    $this->writeBuffer($eq->value, self::OCTETSTRING);
    $this->endSequence();
  }

  /** FILTER_SUBSTRINGS = 0xa4 */
  private function writeFilterSubstring($substr) {
    $this->startSequence(0xa4);
    $this->writeString($substr->attribute);

    $this->startSequence();
    if (null !== $substr->initial) {
      $this->writeString($substr->initial, 0x80);
    }
    foreach ($substr->any as $string) {
      $this->writeString($substr->any, 0x81);
    }
    if (null !== $substr->final) {
      $this->writeString($substr->final, 0x82);
    }
    $this->endSequence();

    $this->endSequence();
  }

  /** FILTER_GE = 0xa5 */
  private function writeFilterGreaterThanEquals($ge) {
    $this->startSequence(0xa5);
    $this->writeString($ge->attribute);
    $this->writeString($ge->value);
    $this->endSequence();
  }

  /** FILTER_LE = 0xa6 */
  private function writeFilterLessThanEquals($le) {
    $this->startSequence(0xa6);
    $this->writeString($le->attribute);
    $this->writeString($le->value);
    $this->endSequence();
  }

  /** FILTER_PRESENT = 0x87 */
  private function writeFilterPresent($present) {
    $this->startSequence(0x87);
    for ($i= 0, $l= strlen($present->attribute); $i < $l; $i++) {
      $this->writeByte($i);
    }
    $this->endSequence();
  }

  /** FILTER_APPROX = 0xa8 */
  private function writeFilterApproximate($approx) {
    $this->startSequence(0xa8);
    $this->writeString($approx->attribute);
    $this->writeString($approx->value);
    $this->endSequence();
  }

  /** FILTER_EXT = 0xa9 */
  private function writeFilterExtensible($approx) {
    $this->startSequence(0xa09);

    $this->writeString($approx->rule, 0x81);
    $this->writeString($approx->type, 0x82);
    $this->writeString($approx->value, 0x83);
    if ($approx->attributes) {
      $this->writeBoolean($approx->type, 0x84);
    }
    $this->endSequence();
  }

  /**
   * Ends current sequences
   *
   * @return void
   */
  public function endSequence() {
    $length= $this->encodeLength(strlen($this->write[0]) - 1);
    $seq= array_shift($this->write);
    $this->write[0].= $seq{0}.$length.substr($seq, 1);
  }

  /**
   * Flushes all sequences to output
   *
   * @return int number of bytes written
   */
  public function flush() {
    $w= $this->out->write($this->write[0]);
    $this->write= [''];
    return $w;
  }

  public function read($l) {
    $chunk= $this->buffer;
    $this->buffer= null;
    while (strlen($chunk) < $l) {
      $bytes= $this->in->read($l - strlen($chunk));
      $this->read[0]-= strlen($bytes);
      $chunk.= $bytes;
    }

    // $t= debug_backtrace();
    // fprintf(STDOUT, "%s   READ %d bytes from %s, remain %d\n", str_repeat('   ', sizeof($this->read)), $l, $t[1]['function'], $this->read[0]);

    return $chunk;
  }

  public function peek() {
    if (null === $this->buffer) {
      $this->buffer= $this->in->read(1);
    }
    return unpack('Ctag', $this->buffer)['tag'];
  }

  public function readTag($expected) {
    $head= unpack('Ctag', $this->read(1));
    $test= (array)$expected;
    if (!in_array($head['tag'], $test)) {
      throw new \lang\IllegalStateException(sprintf(
        'Expected any of [%s], have 0x%02x',
        implode(', ', array_map(function($t) { return sprintf('0x%02x', $t); }, $test)),
        $head['tag']
      ));
    }

    return $head['tag'];
  }

  protected function decodeLength() {
    $head= unpack('Cl0', $this->in->read(1));
    if ($head['l0'] <= 0x7f) {
      $length= $head['l0'];
      $this->read[0]-= 2;
    } else if (0x81 === $head['l0']) {
      $l= unpack('C', $this->in->read(1));
      $length= $l[1];
      $this->read[0]-= 3;
    } else if (0x82 === $head['l0']) {
      $l= unpack('C2', $this->in->read(2));
      $length= $l[1] * 0x100 + $l[2];
      $this->read[0]-= 4;
    } else if (0x83 === $head['l0']) {
      $l= unpack('C3', $this->in->read(3));
      $length= $l[1] * 0x10000 + $l[2] * 0x100 + $l[3];
      $this->read[0]-= 5;
    } else {
      throw new \lang\IllegalStateException('Length too long: '.$head['l0']);
    }
    return $length;
  }

  /**
   * Reads an integer
   *
   * @return int
   */
  public function readInt() {
    $this->readTag(self::INTEGER);
    return unpack('N', str_pad($this->read($this->decodeLength()), 4, "\0", STR_PAD_LEFT))[1];
  }

  /**
   * Reads an enumeration
   *
   * @return int
   */
  public function readEnumeration() {
    $this->readTag(self::ENUMERATION);
    return unpack('N', str_pad($this->read($this->decodeLength()), 4, "\0", STR_PAD_LEFT))[1];
  }

  /**
   * Reads a string
   *
   * @param  int $tag
   * @return string
   */
  public function readString($tag= self::OCTETSTRING) {
    $this->readTag($tag);
    return $this->read($this->decodeLength());
  }

  /**
   * Reads a boolean
   *
   * @return bool
   */
  public function readBoolean() {
    $this->readTag(self::BOOLEAN);
    return $this->read($this->decodeLength()) ? true : false;
  }

  /**
   * Read sequence
   *
   * @param  var $tag expected either a tag or an array of tags
   * @return int The found tag
   */
  public function readSequence($tag= self::SEQ_CTOR) {
    $tag= $this->readTag($tag);
    $len= $this->decodeLength();
    $this->read[0]-= $len;
    array_unshift($this->read, $len);
    // fprintf(STDOUT, "%s`- BEGIN SEQ %d bytes\n", str_repeat('   ', sizeof($this->read)), $this->read[0]);
    return $tag;
  }

  public function remaining() {
    return $this->read[0];
  }

  public function available() {
    return $this->in->available();
  }

  /**
   * Finish reading a sequence
   */
  public function finishSequence() {
    // fprintf(STDOUT, "%s   END SEQ remain: %d bytes\n", str_repeat('   ', sizeof($this->read)), $this->read[0]);
    $shift= array_shift($this->read);
    $this->read[0]-= $shift;
  }

  /**
   * Closes I/O
   *
   * @return void
   */
  public function close() {
    $this->in->close();
    $this->out->close();
  }
}